tests/testthat/helper-functions.R

```diff
@@ -1,419 +1,444 @@
+    stop("Please load RSelenium: library(RSelenium)")
+  if (!"package:testthat" %in% search())
+    stop("Please load testthat: library(testthat)")
+  testDir <- find_test_path(dir)
+  # testthat::test_check assumes we are in path/to/animint/tests
+  old <- getwd()
+  on.exit(setwd(old), add = TRUE)
+  setwd(dirname(testDir))
+  # avoid weird errors if this function is called via testhat::check()
+  # https://github.com/hadley/testthat/issues/144
+  Sys.setenv("R_TESTS" = "")
+  testthat::test_check("animint2", filter = filter)
+}
+
+#' Kill child process(es) that may have been initiated in animint testing
+#'
+#' Read process IDs from a file and kill those process(es)
+#'
+#' @seealso \link{tests_run}
+#' @export
+tests_exit <- function() {
+  res <- stop_binary()
+  Sys.unsetenv("ANIMINT_BROWSER")
+  f <- file.path(find_test_path(), "pids.txt")
+  if (file.exists(f)) {
+    e <- try(readLines(con <- file(f), warn = FALSE), silent = TRUE)
+    if (!inherits(e, "try-error")) {
+      pids <- as.integer(e)
+      res <- c(res, tools::pskill(pids))
+    }
+    close(con)
+    unlink(f)
+  }
+  invisible(all(res))
+}
+
+#' Spawn a child R session that runs a 'blocking' command
+#'
+#' Run a blocking command in a child R session (for example a file server or shiny app)
+#'
+#' @param directory path that the  server should map to.
+#' @param port port number to _attempt_ to run server on.
+#' @param code R code to execute in a child session
+#' @return port number of the successful attempt
+run_servr <- function(directory = ".", port = 4848,
+                      code = "servr::httd(dir='%s', port=%d)") {
+  dir <- normalizePath(directory, winslash = "/", mustWork = TRUE)
+  cmd <- sprintf(
+    paste("write.table(Sys.getpid(), file='%s', append=T, row.name=F, col.names=F);", code),
+    file.path(find_test_path(), "pids.txt"), dir, port
+  )
+  system2("Rscript", c("-e", shQuote(cmd)), wait = FALSE)
+}
+
+# --------------------------
+# Functions that are used in multiple places
+# --------------------------
+
+stop_binary <- function() {
+  if (exists("pJS")) pJS$stop()
+  # these methods are really queries to the server
+  # thus, if it is already shut down, we get some arcane error message
+  e <- try({
+    remDr$closeWindow()
+    remDr$closeServer()
+  }, silent = TRUE)
+  TRUE
+}
+
+# find the path to animint's testthat directory
+find_test_path <- function(dir = ".") {
+  dir <- normalizePath(dir, winslash = "/", mustWork = TRUE)
+  if (!grepl("animint", dir, fixed = TRUE))
+    stop("animint must appear somewhere in 'dir'")
+  base_dir <- basename(dir)
+  if (!base_dir %in% c("animint2", "tests", "testthat"))
+    stop("Basename of dir must be one of: 'animint2', 'tests', 'testhat'")
+  ext_dir <- switch(base_dir,
+                    animint2 = "tests/testthat",
+                    tests = "testthat",
+                    testthat = "")
+  file.path(dir, ext_dir)
+}
+
+runtime_evaluate <- function(script=NULL,return.value=FALSE){
```

Comment:
it looks like return.value=FALSE is default, but never used in the test code. If that is true, then please remove return.value arg, and delete the FALSE branch. If not, can you please explain why that branch is needed?

Reply:
Am I implementing your suggestion in https://github.com/animint/animint2/pull/126#discussion_r1677127828 correctly ? return.value=FALSE by default, so the function will return eval.result, if return.value becomes TRUE then the function will return eval.result$result$value

Reply:
right, but I'm just wondering if we need to support return.value=FALSE at all, because I don't see it used anywhere, do you?

Reply:
> right, but I'm just wondering if we need to support return.value=FALSE at all, because I don't see it used anywhere, do you?

@tdhock I am using return.value in if-else condition.

to give an example, in line 81 of knit-print test I pass return.value=TRUE to runtime_evaluate_helper like this:

a<runtime_evaluate_helper(id=as.character(id),class_name='show_hide_selector_widgets',list_num=0,return.value=TRUE)

 that value is propagated by runtime_evaluate_helper to runtime_evaluate.
 
so runtime_evaluate will now return by value. if no return.value is passed default value of return.value is FALSE. so it won't return by value. it will just return eval.result

Reply:
> right, but I'm just wondering if we need to support return.value=FALSE at all, because I don't see it used anywhere, do you?

I think now I know what you mea. So I can just keep return.value=NULL instead of return.value=FALSE ?

Reply:
no I am not suggesting to keep return.value=NULL.
I am wondering if we can remove support for return.value=FALSE?
I see you cited line 81 of knit-print test where return.value=TRUE is used.
is return.value=FALSE ever used? If not then we can delete that code as it is un-necessary. (potentially confusing)

Reply:
> no I am not suggesting to keep return.value=NULL. I am wondering if we can remove support for return.value=FALSE? I see you cited line 81 of knit-print test where return.value=TRUE is used. is return.value=FALSE ever used? If not then we can delete that code as it is un-necessary. (potentially confusing)

@tdhock 

return.value is used in the following if-else condition inside runtime_evaluate function :

 if (return.value){
    eval.result$result$value
  }else{
    eval.result
  }

I am using the if-else mentioned above because Runtime$evaluate(script,returnByValue = TRUE) is always returning by value, which is then stored in eval.result

Reply:
OK I will try to answer my own question. I used grep to find all the uses of runtime_evaluate_helper.
```
tests/testthat/helper-functions.R:  runtime_evaluate_helper(id=v,dispatch_event=TRUE)
tests/testthat/test-renderer1-knit-print.R:    runtime_evaluate_helper(id=as.character("plot1top"),class_name='show_hide_selector_widgets',list_num=0,dispatch_event=TRUE)
tests/testthat/test-renderer1-knit-print.R:    runtime_evaluate_helper(id=as.character("plot1top"),class_name='selectize-input',list_num=0,dispatch_event=TRUE)
tests/testthat/test-renderer1-knit-print.R:    runtime_evaluate_helper(id=as.character("plot1bottom"),class_name='show_hide_selector_widgets',list_num=0,dispatch_event=TRUE)
tests/testthat/test-renderer1-knit-print.R:    runtime_evaluate_helper(id=as.character("plot1bottom"),class_name='selectize-input',list_num=0,dispatch_event=TRUE)
tests/testthat/test-renderer2-widerect.R:runtime_evaluate_helper(class_name="show_hide_selector_widgets",list_num=0,dispatch_event=TRUE)
```
Below are with return.value=TRUE, and above are without. Below we see the result saved in a variable, for example `a` or `b` or `show_hide`. Above we see that the result is not saved in a variable, so that suggests to me that we should simplify the code: delete the `return.value` argument, and always return `eval.result$result$value`. Does that make sense to you? Or do you see some reason why we need to keep returning the `eval.result` ? (even though it is not used in the uses above)
```
tests/testthat/test-renderer1-knit-print.R:  a<- runtime_evaluate_helper(id=as.character(id),class_name='show_hide_selector_widgets',list_num=0,return.value=TRUE)
tests/testthat/test-renderer1-knit-print.R:  b<- runtime_evaluate_helper(id=as.character(id),class_name='show_hide_selector_widgets',list_num=0,return.value=TRUE)
tests/testthat/test-renderer1-knit-print.R:  show_hide <- runtime_evaluate_helper(id=as.character(id),class_name='table.legend tr.label_variable',list_num=0,return.value=TRUE)
tests/testthat/test-renderer1-knit-print.R:  widget <- runtime_evaluate_helper(id=as.character(id),class_name='table.legend tr.label_variable',list_num=1,return.value=TRUE)
```
also please fix
* change single letter variable names like `a` and `b` to informative variable names like `show_hide`
* change as.character("plot1top") to "plot1top" to simplify (as.character is not necessary)

Reply:
@tdhock I was focusing only on the helper function. thank you for pointing out why we can always return result by value even the call does not explicitly want that. I will make the change

Reply:
this was done